TimeCalibration/TimeCalibration.cpp

```diff
@@ -0,0 +1,454 @@
+	  output.close();
+	  }
+	 //
+		for (int thr=1;thr<=4;thr++){// loop over thresholds
+		  
+//histos for leading edge
+//		  const char * histo_name_l = formatUniqueSlotDescription(scin.at()->getBarrelSlot(), thr, "timeDiffAB_leading_");
+//
+		        const char * histo_name_l = Form("%slayer_%d_slot_%d_thr_%d","timeDiffAB_leading_",LayerToCalib,StripToCalib,thr);
+			getStatistics().createHistogram( new TH1F(histo_name_l, histo_name_l, 400, -20., 20.) );
+			//
+//histograms for leading edge refference detector time difference
+                        const char * histo_name_Ref_l = Form("%slayer_%d_slot_%d_thr_%d","timeDiffRef_leading_",LayerToCalib,StripToCalib,thr);
+			getStatistics().createHistogram( new TH1F(histo_name_Ref_l, histo_name_Ref_l, 800,-80.,80.) );
+			//
+//histos for trailing edge			
+			const char * histo_name_t = Form("%slayer_%d_slot_%d_thr_%d","timeDiffAB_trailing_",LayerToCalib,StripToCalib,thr);
+			getStatistics().createHistogram( new TH1F(histo_name_t, histo_name_t, 400, -20., 20.) );
+			//
+//histograms for leading edge refference detector time difference
+                        const char * histo_name_Ref_t = Form("%slayer_%d_slot_%d_thr_%d","timeDiffRef_trailing_",LayerToCalib,StripToCalib,thr);
+			getStatistics().createHistogram( new TH1F(histo_name_Ref_t, histo_name_Ref_t, 1000,-100.,100.) );
+			//
+		}
+ INFO("#############");       
+ INFO("CALIB_INIT: INITIALIZATION DONE!");       
+ INFO("#############");
+}
+
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void TimeCalibration::exec(){
+  double RefTimeLead[4]={-1.e43,-1.e43,-1.e43,-1.e43};
+  double RefTimeTrail[4]={-1.e43,-1.e43,-1.e43,-1.e43};
+  std::vector <JPetHit> fhitsCalib;
+  std::vector <double> fRefTimesL;
+  std::vector <double> fRefTimesT;
+
+  const int kPMidRef = 385;  
+  //getting the data from event in propriate format
+  if(auto timeWindow = dynamic_cast<const JPetTimeWindow* const>(getEvent())) {
+    uint n = timeWindow->getNumberOfEvents();
+    //
+    for(uint i=0;i<n;++i){
+      const JPetHit & hit = dynamic_cast<const JPetHit&>(timeWindow->operator[](i));
+	  int PMid = hit.getSignalB().getRecoSignal().getRawSignal().getPM().getID();
+         //
+         //taking refference detector hits times (scin=193, Pmt=385)
+         //
+
+	  if(PMid==kPMidRef){
+	    auto lead_times_B = hit.getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Leading);
+	    auto trail_times_B = hit.getSignalB().getRecoSignal().getRawSignal().getTimesVsThresholdNumber(JPetSigCh::Trailing);                                                   
+	    //
+	    for(auto & thr_time_pair : lead_times_B){
+	      int thr = thr_time_pair.first;
+	      RefTimeLead[thr] = thr_time_pair.second;

```

Comment:
As already mentioned by Kamil, you have defined arrays RefTimeLead with four elements with indice 0-3, while ThresholdNumbers are numbered from 1-4. This may cause the situatuon in which thr =4
and you call RefTimeLead[4]  which is a severe bug.
This should be cross-checked and fixed. ( @alekgajos @mszonk @grey277)

Reply:
Done